code/modules/lighting/lighting_turf.dm

```diff
@@ -1,26 +1,76 @@
 /turf
-	var/list/affecting_lights
-	var/atom/movable/lighting_overlay/lighting_overlay
+	var/dynamic_lighting = 1
+	luminosity           = 1
 
+	var/tmp/lighting_corners_initialised = FALSE
+
+	var/tmp/list/datum/light_source/affecting_lights       // List of light sources affecting this turf.
+	var/tmp/atom/movable/lighting_overlay/lighting_overlay // Our lighting overlay.
+	var/tmp/list/datum/lighting_corner/corners[4]
+	var/tmp/has_opaque_atom = FALSE // Not to be confused with opacity, this will be TRUE if there's any opaque atom on the tile.
+
+/turf/New()
+	. = ..()
+
+	if (opacity)
+		has_opaque_atom = TRUE
 /turf/proc/reconsider_lights()
 	for(var/datum/light_source/L in affecting_lights)
 		L.vis_update()
 
-/turf/proc/lighting_clear_overlays()
+/turf/proc/lighting_clear_overlay()
 	if(lighting_overlay)
 		qdel(lighting_overlay)
 
-/turf/proc/lighting_build_overlays()
+/turf/proc/lighting_build_overlay()
 	if(lighting_overlay)
 		return
 
 	var/area/A = loc
-	if(A.lighting_use_dynamic)
+	if(A.dynamic_lighting)
 		var/atom/movable/lighting_overlay/O = new(src)
 		lighting_overlay = O
-		// forcibly refresh lighting on this tile
-		// must make light forget tile or it wil remain
-		// convinced that it is already lighting it
-		for(var/datum/light_source/L in affecting_lights)
-			L.forget_turf(src)
-			L.vis_update()
\ No newline at end of file
+
+		for (var/datum/lighting_corner/C in corners)
+			if (!C.active) // We would activate the corner, calculate the lighting for it.
+				for (var/L in C.affecting)
+					var/datum/light_source/S = L
+					S.recalc_corner(C)
+
+				C.active = TRUE
+
+// Can't think of a good name, this proc will recalculate the has_opaque_atom variable.
+/turf/proc/recalc_atom_opacity()
+	has_opaque_atom = FALSE
+	for (var/atom/A in src.contents + src) // Loop through every movable atom on our tile PLUS ourselves (we matter too...)
+		if (A.opacity)
+			has_opaque_atom = TRUE
+
+// If an opaque movable atom moves around we need to potentially update visibility.
+/turf/Entered(var/atom/movable/Obj, var/atom/OldLoc)
```

Comment:
As long as Exited still gets called it should be fine.

Still pretty crap it doesn't call the parent.